gk/fib.c

```diff
@@ -0,0 +1,1343 @@
+		if (fib_id < 0)
+			goto out;
+
+		new_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Add the fib entry for the IPv6 address 
+		 * to the IPv6 LPM table.
+		 */
+		rte_memcpy(ipv6_route.ip,
+			ip_addr->ip.v6.s6_addr, sizeof(ipv6_route.ip));
+		ipv6_route.depth = prefix_len;
+		ipv6_route.nexthop = fib_id;
+		ret = gk_lpm_add_ipv6_routes(&ipv6_route, 1, ltbl);
+		if (ret < 0) {
+			new_fib = NULL;
+			goto out;
+		}
+	} else
+		goto out;
+
+	new_fib->ref_cnt = 1;
+
+out:
+	return new_fib;
+}
+
+static int
+setup_neighbor_tbl(unsigned int socket_id, int identifier,
+	int ip_ver, int ht_size, struct neighbor_hash_table *neigh)
+{
+	int  ret;
+	char ht_name[64];
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+		sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	struct rte_hash_parameters neigh_hash_params = {
+		.entries = ht_size,
+		.key_len = key_len,
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+	};
+
+	ret = snprintf(ht_name, sizeof(ht_name),
+		"neighbor_hash_%u", identifier);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(ht_name));
+
+	neigh->tbl_size = ht_size;
+
+	/* Setup the neighbor hash table. */
+	neigh_hash_params.name = ht_name;
+	neigh_hash_params.socket_id = socket_id;
+	neigh->hash_table = rte_hash_create(&neigh_hash_params);
+	if (neigh->hash_table == NULL) {
+		RTE_LOG(ERR, HASH,
+			"The GK block cannot create hash table for neighbor FIB!\n");
+
+		ret = -1;
+		goto out;
+	}
+
+	/* Setup the Ethernet header cache table. */
+	neigh->cache_tbl = rte_calloc(NULL,
+		ht_size, sizeof(struct ether_cache), 0);
+	if (neigh->cache_tbl == NULL) {
+		RTE_LOG(ERR, MALLOC,
+			"The GK block cannot create Ethernet header cache table\n");
+
+		ret = -1;
+		goto neigh_hash;
+	}
+
+	ret = 0;
+	goto out;
+
+neigh_hash:
+	rte_hash_free(neigh->hash_table);
+	neigh->hash_table = NULL;
+out:
+	return ret;
+}
+
+/*
+ * Setup the FIB entries for the network prefixes, for which @iface
+ * is responsible.
+ * These prefixes are configured when the Gatekeeper server starts.
+ */
+static int
+setup_net_prefix_fib(int identifier,
+	struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int ret;
+	int ip_prefix_len;
+	struct ipaddr addr;
+	unsigned int socket_id = rte_lcore_to_socket_id(gk_conf->lcores[0]);
+	struct net_config *net_conf = gk_conf->net;
+	struct gk_fib *fib[2] = { NULL, NULL };
+
+	/* Set up the FIB entry for the IPv4 network prefix. */
+	if (ipv4_if_configured(iface)) {
+		addr.proto = ETHER_TYPE_IPv4;
+		rte_memcpy(&addr.ip.v4, &iface->ip4_addr, sizeof(addr.ip.v4));
+		ip_prefix_len = iface->ip4_addr_plen;
+
+		fib[0] = add_prefix_fib(&addr, ip_prefix_len, gk_conf);
+		if (fib[0] == NULL)
+			return -1;
+
+		if (iface == &net_conf->front) {
+			fib[0]->action = GK_FWD_NEIGHBOR_FRONT_NET;
+			gk_conf->neigh_fib_front = fib[0];
+		} else if (iface == &net_conf->back) {
+			fib[0]->action = GK_FWD_NEIGHBOR_BACK_NET;
+			gk_conf->neigh_fib_back = fib[0];
+		} else
+			return -1;
```

Comment:
Yes, the code shouldn't reach here, since we only setup FIB entries for either ``front`` or ``back`` interface. So I added a ``rte_panic()``.